### PR TITLE
Add covariant return types to methods in MultiReaderList that return `this`.

### PR DESCRIPTION
**Pull request by motlin**

### CODE REVIEW

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MultiReaderList.java [hunk 1]
```diff
@@ -23,4 +36,65 @@
...
+    @Override
+    MultiReaderList<T> sortThisByLong(LongFunction<? super T> function);
+
+    @Override
+    MultiReaderList<T> sortThisByDouble(DoubleFunction<? super T> function);
+
+    @Override
+    default MultiReaderList<T> reverseThis()
+    {
+        Collections.reverse(this);
```
Comment (donraab) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MultiReaderList.java:83]: This needs to be protected by a call to `withWriteLockAndDelegate`.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MultiReaderList.java [hunk 2]
```diff
@@ -23,4 +36,65 @@
...
+        return this;
+    }
+
+    @Override
+    default MultiReaderList<T> shuffleThis()
+    {
+        Collections.shuffle(this);
```
Comment (donraab) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MultiReaderList.java:90]: Same as above.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MultiReaderList.java [hunk 3]
```diff
@@ -23,4 +36,65 @@
...
+        return this;
+    }
+
+    @Override
+    default MultiReaderList<T> shuffleThis(Random random)
+    {
+        Collections.shuffle(this, random);
```
Comment (donraab) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/list/MultiReaderList.java:97]: Same as above.

**Comment by mohrezaei** [comment]
As a more general rule to the "this needs to be protected by XYZ lock", I would suggest any delegating collection we have ought to always call the appropriate method on the delegate, instead of an external implementation, such as `Collections`. Said differently, not seeing the delegation should be a code smell.

**Comment by motlin** [comment]
Sorry, this was meant to be a WIP PR. It didn't even compile.

There are no default methods now. This is a case where it's important for the overridden interface methods to be abstract.

We can reduce duplication more by pulling up default method bodies that delegate to withReadLockAndDelegate() and withWriteLockAndDelegate().

**Pull request merged**